keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,522 @@
+
+#### Story 1
+
+Consider a cloud provider where servers are cheaper in the night. My Job takes
+several days to complete, so I'd like to stop my Job early in the morning every
+day and resume it after dark to save money. However, I don't want to delete and
+re-create my Job every day as I don't want to lose track of completed Pods,
+logs, and other metadata.
+
+#### Story 2
+
+I'm the system administrator of a cluster than can only run one Job at any
+given time. I'd like my users to submit their Jobs whenever they want, but I
+want control over whose Job gets to run on the cluster. With a custom mutating
+webhook that adds `stopped: true` to all user-submitted Jobs, I can
+indefinitely delay Pod creation.
+
+### Notes/Constraints/Caveats
+
+System administrators may want to restrict who can set/modify the `stopped`
+field when creating or updating Jobs. This can be achieved with validating
+and/or mutating webhooks.
+
+When a Job is stopped, the time will be recorded as a part of the Job status.
+Users can infer how long a Job has been paused for with this field. This can be
+useful when making decisions around which Job should be resumed.
+
+A Job that is complete cannot be stopped.
+
+The feature will be gated behind a feature gate.
+
+### Risks and Mitigations
+
+1. Stopping an active Job will delete *all* active pods. Users must design
+   their application to gracefully handle this.
+
+## Design Details
+
+We propose adding a `stopped` field to the JobSpec API:
+
+```go
+type JobSpec struct {
+	// Stopped specifies whether the Job controller should create Pods or not. If
+	// a Job is created in the stopped state, no Pods are created by the Job
+	// controller. If a Job enters the stopped state (i.e. the flag goes from
+	// false to true), the Job controller will delete all Pods associated with
+	// this Job. This is an alpha field and requires enabling the StoppedJob
+	// feature gate. Defaults to false.
+	Stopped *bool `json:"stopped,omitempty"`
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager will abstain from creating Pods even if the Job's
+parallelism requirements aren't met. If the Job is already active and is
+updated with `stopped: true`, all Pods owned by this Job will be deleted.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition called "Stopped"
+that will be a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+type JobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// JobStopped means the job has been stopped.
+	JobStopped JobConditionType = "Stopped"
```

Comment:
A couple of questions:
1) should the job transition to another new condition when the job is unstopped? 
2) will this condition exist in case the job was created in stopped state from the get go?

Reply:
>should the job transition to another new condition when the job is unstopped?

I don't think so. There is currently no status when a job is active. I think a resumed job should be as similar to a never-stopped job as possible.

Besides, the condition should be statement about the present, not the past ("the job *was* unstopped, the job *is* active/stopped/complete").

>will this condition exist in case the job was created in stopped state from the get go?

Yes.

Reply:
the other two conditions are terminal, as in once the job reaches them it will never transition out. The stopped one is different in that you can transition out of it. So if we are not transitioning to another condition after the job has been resumed, then a user can't rely on the job condition to check whether or not the job is stopped, it should always check the flag. Which brings two questions: what is the value of having a stopped JobCondition?

Reply:
>a user can't rely on the job condition to check whether or not the job is stopped

The presence/lack of the Stopped condition will tell them if the job is stopped or not.

>what is the value of having a stopped JobCondition

There is a non-zero amount of time between updating the Job spec and the Job controller observing the change. If the native k8s Job controller is buggy or down, a Job could have `stopped: true` without a Stopped condition indefinitely.

The controller will add the Stopped condition only *after* it has requested the API server to delete all pods.

---

So JobConditions also have a boolean Status field. We could transition from `Stopped: true` to `Stopped: false` in the status when unstopped. Jobs that have never been stopped will not have the condition at all. Updated the KEP to reflect this.

Reply:
ok, so we are transitioning to another state: condition.Status=false when unstopping the job, this clarified it. thanks.